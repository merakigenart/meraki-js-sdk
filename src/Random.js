export class Random {
    prngA;
    prngB;

    constructor(tokenData = { tokenHash: '', tokenId: '' }) {
        this.useA = false;
        let sfc32 = class {
            handler;
            constructor(uint128Hex) {
                let a = parseInt(uint128Hex.substr(0, 8), 16);
                let b = parseInt(uint128Hex.substr(8, 8), 16);
                let c = parseInt(uint128Hex.substr(16, 8), 16);
                let d = parseInt(uint128Hex.substr(24, 8), 16);

                this.handler = function () {
                    a |= 0;
                    b |= 0;
                    c |= 0;
                    d |= 0;
                    let t = (((a + b) | 0) + d) | 0;
                    d = (d + 1) | 0;
                    a = b ^ (b >>> 9);
                    b = (c + (c << 3)) | 0;
                    c = (c << 21) | (c >>> 11);
                    c = (c + t) | 0;
                    return (t >>> 0) / 4294967296;
                };
            }
        };
        const hashPartLength = (tokenData.tokenHash.length - 2) / 2;
        console.log('hashPartLength', hashPartLength);

        // seed prngA with first half of tokenData.hash
        this.prngA = new sfc32(tokenData.tokenHash.substr(2, hashPartLength)).handler;
        // seed prngB with second half of tokenData.hash
        this.prngB = new sfc32(tokenData.tokenHash.substr(hashPartLength + 2, hashPartLength)).handler;

        for (let i = 0; i < 1e6; i += 2) {
            this.prngA();
            this.prngB();
        }
    }

    // random number between 0 (inclusive) and 1 (exclusive)
    decimal() {
        this.useA = !this.useA;
        return this.useA ? this.prngA() : this.prngB();
    }

    // random number between a (inclusive) and b (exclusive)
    number(a, b) {
        if (a === undefined) {
            a = 0;
        }

        if (b === undefined) {
            b = Number.MAX_VALUE;
        }

        return a + (b - a) * this.decimal();
    }

    // random integer between a (inclusive) and b (inclusive)
    // requires a < b for proper probability distribution
    integer(a, b) {
        if (a === undefined) {
            a = 0;
        }

        if (b === undefined) {
            b = Number.MAX_VALUE;
        }

        return Math.floor(this.number(a, b + 1));
    }

    // random boolean with p as percent likelihood of true
    boolean(p = 50) {
        return this.decimal() < p * 0.1;
    }

    // random value in an array of items
    element(list) {
        return list[this.integer(0, list.length - 1)];
    }
}
