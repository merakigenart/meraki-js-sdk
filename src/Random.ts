// @ts-nocheck

//import { generateRandomTokenData } from './helpers';

export class Random {
    protected state: Record<string, any>;
    protected seedValues: Record<string, any>;

    prngA;
    prngB;
    useA = false;

    constructor(public tokenData = { tokenHash: '', tokenId: '' }) {
        this.useA = false;
        const sfc32 = class {
            handler;
            constructor(uint128Hex) {
                let a = parseInt(uint128Hex.substr(0, 8), 16);
                let b = parseInt(uint128Hex.substr(8, 8), 16);
                let c = parseInt(uint128Hex.substr(16, 8), 16);
                let d = parseInt(uint128Hex.substr(24, 8), 16);

                this.handler = function (seed = 4294967296) {
                    if (seed === undefined) {
                        seed = 4294967296;
                    }

                    a |= 0;
                    b |= 0;
                    c |= 0;
                    d |= 0;
                    const t = (((a + b) | 0) + d) | 0;
                    d = (d + 1) | 0;
                    a = b ^ (b >>> 9);
                    b = (c + (c << 3)) | 0;
                    c = (c << 21) | (c >>> 11);
                    c = (c + t) | 0;
                    return (t >>> 0) / seed;
                };
            }
        };

        const hashPartLength = (tokenData.tokenHash.length - 2) / 2;

        // const seedStr = tokenData.tokenHash.slice(2);
        // const seeds = [];

        // for(let i = 0; i <= 64; i++) {
        //     seeds.push(seedStr.substring(i, 4));
        //     i += 4;
        // }

        // seed prngA with first half of tokenData.hash
        this.prngA = new sfc32(tokenData.tokenHash.substring(2, hashPartLength)).handler;
        // seed prngB with second half of tokenData.hash
        this.prngB = new sfc32(tokenData.tokenHash.substring(hashPartLength + 2, hashPartLength)).handler;

        for (let i = 0; i < 1e6; i += 2) {
            this.prngA();
            this.prngB();
        }

        //this.tokenData.tokenHash = generateRandomTokenData(2).tokenHash;

        this.state = this.initializeState().state;
        this.seedValues = this.initializeSeeds(this.generateSeeds(this.tokenData.tokenHash));
        this.rnd();
    }

    // random number between 0 (inclusive) and 1 (exclusive)
    decimal() {
        return this.rnd();
    }

    // random number between a (inclusive) and b (exclusive)
    number(a = undefined, b = undefined) {
        if (a === undefined) {
            a = 0;
        }

        if (b === undefined) {
            b = Number.MAX_VALUE - 2;
        }

        return a + (b - a) * this.decimal();
    }

    // random integer between a (inclusive) and b (inclusive)
    // requires a < b for proper probability distribution
    integer(a = undefined, b = undefined) {
        if (a === undefined) {
            a = 0;
        }

        if (b === undefined) {
            b = Number.MAX_VALUE - 2;
        }

        return Math.floor(this.number(a, b + 1));
    }

    // random boolean with p as percent likelihood of true
    boolean(p = 50) {
        return this.decimal() < p * 0.1;
    }

    // random value in an array of items
    element(list) {
        return list[this.integer(0, list.length - 1)];
    }

    generateSeeds(str: string): number[] {
        let part = 0;
        const seeds: number[] = [];

        if (str.startsWith('0x')) {
            str = str.slice(2);
        }

        for (let i = 0; i < str.length; i++) {
            part = str.slice(i, i + 4);
            seeds.push(parseInt(part, 16));

            i += 4;
        }

        for (let i = 0; i < str.length; i++) {
            part = str.slice(i, i + 4);
            seeds.push(parseInt(part, 16));

            i += 2;
        }

        for (let i = 0; i < str.length; i++) {
            part = str.slice(i, i + 4);
            seeds.push(parseInt(part, 16));

            i += 1;
        }

        for (let i = 0; i < str.length; i++) {
            part = str.substring(str.length - i - 4, str.length - i);
            part = part.substring(2, 4) + part.substring(0, 2);
            seeds.push(parseInt(part, 16));
            i += 4;
        }

        return seeds;
    }

    protected initializeSeeds(seeds: number[]) {
        const seedValues = {
            eps: Math.pow(2, -32),
            m0: seeds.shift(),
            m1: seeds.shift(),
            m2: seeds.shift(),
            m3: seeds.shift(),
            a0: seeds.shift(),
            a1: seeds.shift(),
            a2: seeds.shift(),
            a3: seeds.shift(),
        };

        return seedValues;
    }

    protected initializeState(stateSize = 4, integerSize: 8 | 16 | 32 | 64 = 16) {
        const intMap = {
            8: Uint8Array,
            16: Uint16Array,
            32: Uint32Array,
            64: BigUint64Array,
        };

        const intClass = intMap[integerSize];
        const state = new intClass(stateSize);
        const dataView = new DataView(state.buffer);

        return {
            integerSize,
            stateSize,
            state,
            dataView,
        };
    }

    rnd() {
        const { eps, a0, a1, a2, a3, m0, m1, m2, m3 } = this.seedValues;

        const a = this.state[0],
            b = this.state[1],
            c = this.state[2],
            e = this.state[3],
            f = 0 | (a0 + m0 * a),
            g = 0 | (a1 + m0 * b + (m1 * a + (f >>> 16))),
            h = 0 | (a2 + m0 * c + m1 * b + (m2 * a + (g >>> 16)));

        this.state[0] = f;
        this.state[1] = g;
        this.state[2] = h;
        this.state[3] = a3 + m0 * e + (m1 * c + m2 * b) + (m3 * a + (h >>> 16));

        const i = (e << 21) + (((e >> 2) ^ c) << 5) + (((c >> 2) ^ b) >> 11);

        return eps * (((i >>> (e >> 11)) | (i << (31 & -(e >> 11)))) >>> 0);
    }
}
