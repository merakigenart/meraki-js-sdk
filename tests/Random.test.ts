/* eslint-disable no-undef */

import { Random } from '@/Random';

let random: Random;

beforeEach(() => {
    random = new Random({ tokenId: '123', tokenHash: '123' });
});

it('generates a random integer', () => {
    expect(random.integer()).toBeGreaterThanOrEqual(0);
});

it('generates a random decimal', () => {
    expect(random.decimal()).toBeGreaterThanOrEqual(0.0);
});

it('generates a random boolean', () => {
    expect(typeof random.boolean()).toBe('boolean');
});

it('generates random decimals', () => {
    const decimal1 = random.decimal(123456);
    const decimal2 = random.decimal(3456789);

    expect(decimal1).not.toEqual(decimal2);
});
