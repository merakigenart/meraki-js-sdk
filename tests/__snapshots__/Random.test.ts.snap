// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates predictable seeds based on a provided value 1`] = `
[
  37900,
  42791,
  17987,
  9711,
  36119,
  4555,
  115,
  37900,
  52391,
  29300,
  17508,
  17186,
  9711,
  61581,
  53623,
  28956,
  52103,
  115,
  37900,
  3274,
  51826,
  29300,
  29766,
  17987,
  17186,
  8798,
  24304,
  61581,
  36119,
  6001,
  28956,
  7352,
  47219,
  115,
  29624,
  4471,
  36336,
  9522,
  18036,
  42956,
  148,
]
`;

exports[`generates predictable seeds based on a provided value 2`] = `
[
  37900,
  42791,
  17987,
  9711,
  36119,
  4555,
  115,
  37900,
  52391,
  29300,
  17508,
  17186,
  9711,
  61581,
  53623,
  28956,
  52103,
  115,
  37900,
  3274,
  51826,
  29300,
  29766,
  17987,
  17186,
  8798,
  24304,
  61581,
  36119,
  6001,
  28956,
  7352,
  47219,
  115,
  29624,
  4471,
  36336,
  9522,
  18036,
  42956,
  148,
]
`;

exports[`generates predictable seeds based on a provided value 3`] = `
[
  26528,
  59619,
  267,
  13106,
  5197,
  29244,
  183,
  26528,
  1000,
  36407,
  28688,
  3043,
  13106,
  10260,
  17628,
  50979,
  15467,
  183,
  26528,
  41022,
  16014,
  36407,
  14081,
  267,
  3043,
  58163,
  13096,
  10260,
  5197,
  19911,
  50979,
  9158,
  50871,
  183,
  47046,
  29404,
  5160,
  13246,
  311,
  59395,
  103,
]
`;

exports[`generates predictable seeds based on a provided value 4`] = `
[
  4660,
  26505,
  4660,
  17767,
  30864,
  0,
  4660,
  13398,
  22136,
  30864,
  144,
  36984,
  17699,
]
`;
